third_party/pathway-adaptive-rag.ipynb

```diff
@@ -0,0 +1,378 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "268b24c1-24da-4fc9-a65e-229e8132db59",
+   "metadata": {},
+   "source": [
+    "# Adaptive RAG with Pathway\n",
+    "\n",
+    "This notebook shows how you can **dynamically adapt the number of documents in a RAG prompt** using feedback from the LLM. This can give you significant cost reduction of RAG LLM question answering pipelines while maintaining good accuracy. Mistral makes it easy to run Adaptive RAG locally with `ollama` or remotely using the Mistral API.\n",
+    "\n",
+    "Reference paper: https://pathway.com/developers/showcases/adaptive-rag\n",
+    "\n",
+    "<img src=\"images/pw-adaptive-rag.png\" width=\"700\"/>\n",
+    "\n",
+    "Let's jump in! 🪂\n",
+    "\n",
+    "## Setup\n",
+    "\n",
+    "### Using APIs \n",
+    "\n",
+    "* You will need [a Mistral subscription](https://auth.mistral.ai/ui/login?flow=2b98deac-f13f-4e18-a7cd-23ba377a6370) to access the Mistral API.\n",
+    "* Create an account and fetch your API Key\n",
+    "* Pass the API Key into the prompt in this notebook or set it as environment variable `MISTRAL_API_KEY`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4a6a2ffd-c3ce-4cb7-a7a5-b32d098807eb",
+   "metadata": {},
+   "source": [
+    "### Installing Libraries\n",
+    "You will need to install `pathway` to run this code. We will also install `litellm` and `sentence-transformers` which are optional dependencies of Pathway."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "860c99fb",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# # Uncomment and run if you need to install Pathway and Mistral packages\n",
+    "# !pip install -U --prefer-binary pathway~= 0.9.0 \n",
```

Comment:
I believe "pathway~=0.9.0" has to be in between ", also run_local needs to be declared even if It's False.